### PR TITLE
server: Automatically repair redis AOF

### PR DESCRIPTION
**Pull request by keegancsmith**
If the AOF file exists, we automatically run the repair command on it. We do it in a best-effort fashion. So if the repair fails we will still try start redis, but will include in the log the failure. This should help a common reason our server container won't start. For example this has happened to sourcegraph.sgdev.org twice in the last week.

Test plan: CI. Haven't tested this closely at all, but will rely on CI that the image isn't broken + best effort nature to know we haven't broken anything in the bad case.

Fixes https://github.com/sourcegraph/sourcegraph/issues/651

### CODE REVIEW

File: cmd/server/shared/redis.go [hunk 1]
```diff
@@ -60,10 +62,39 @@ func maybeRedisProcFile(c redisProcfileConfig) (string, error) {
...
 func tryCreateRedisConf(c redisProcfileConfig) (string, error) {
+	dataDir := filepath.Join(os.Getenv("DATA_DIR"), c.dataDir)
+	err := os.MkdirAll(dataDir, os.FileMode(0755))
+	if err != nil {
+		return "", err
+	}
+
+	// Best-effort repair AOF in case it is corrupted
+	// https://github.com/sourcegraph/sourcegraph/issues/651
+	aofPath := filepath.Join(dataDir, "appendonly.aof")
```
Comment (tsenart) [inline cmd/server/shared/redis.go]: Is this the actual file name on disk? What I wrote in the issue was just exemplary.

Comment (uwedeportivo) [reply cmd/server/shared/redis.go]: Is AOF part of conf ? Maybe we should extract it into a function tryFixRedisAOF if it's not.

Comment (keegancsmith) [reply cmd/server/shared/redis.go]: after some googling... looks like I should test this. For example the script is interactive.

**Comment by slimsag** [comment]
Is there no way we can do this in our redis images themselves? It seems weird that we would have `sourcegraph/server` be self-correcting in this regard, but our actual HA cluster deployments would not be?

**Comment by tsenart** [comment]
> Is there no way we can do this in our redis images themselves? It seems weird that we would have sourcegraph/server be self-correcting in this regard, but our actual HA cluster deployments would not be?

Yes, but that would not solve the problem in the server image, which I believe is where this happens most of the time because of all the different processes that it contains in a single container causing sporadic one-for-all abrupt (non-graceful) terminations.

**Comment by keegancsmith** [comment]
> Is there no way we can do this in our redis images themselves? It seems weird that we would have sourcegraph/server be self-correcting in this regard, but our actual HA cluster deployments would not be?

We can, but it will still require us doing it here. As for HA cluster this is less likely to occur since Redis being shutdown won't happen due to frontend constantly crashing/etc. It will shutdown only because k8s tells it to shutdown, and will have a reasonable time to shutdown. But I have seen it happen in the case of the Redis image constantly ooming, but that required admin intervention anyways. Essentially it is much more common for server than it should be for HA. WDYT?

**Comment by slimsag** [comment]
Ah, I see, that is a really good point I missed. Sorry! This approach SGTM, then. If we see this occurring in clusters, then we can address it there seperately if needed.

**Comment by keegancsmith** [comment]
Behold:
- E2E testing of corrupted Redis AOF files! 
- The "yes" command as an `io.Reader`.

**Pull request merged**